### PR TITLE
Add MaxRL mean normalization over advantages

### LINKED ISSUE #1030: [algorithm] Add MaxRL mean normalization over advantages
MaxRL:  https://zanette-labs.github.io/MaxRL/

Add mean normalization to advantages as an option to implement MaxRL

Image

### PR DESCRIPTION
**Pull request by tamoghnokandar**
Fixes #1030 

---

<a href="https://app.devin.ai/review/novasky-ai/skyrl/pull/1126" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

**Comment by SumanthRH** [comment]
Hi @tamoghnokandar .Thank you for the PR! We recently finished the migration from `skyrl-train` + `skyrl-tx` -> `skyrl` and are just now getting to open PRs! 

 Is this PR ready for review? Could you merge the latest changes from `main` if so?

**Comment by tamoghnokandar** [comment]
Hi @SumanthRH , I merged the latest changes from main. Ready for review now.

File: tests/tx/models/test_models_common.py
Comment (SumanthRH) [inline tests/tx/models/test_models_common.py:1]: why was this file modified? can you revert these changes?

File: tests/backends/skyrl_train/utils/test_ppo_utils.py [hunk 1]
```diff
@@ -172,6 +173,32 @@ def test_compute_grpo_outcome_advantage_norm_std_false():
...
+    response_mask = torch.ones_like(token_level_rewards)
+    index = np.array([0, 0, 1, 1])
+
+    adv, ret = compute_maxrl_advantage(
+        token_level_rewards=token_level_rewards,
+        response_mask=response_mask,
+        index=index,
+    )
+
+    expected = torch.tensor([1.5 / 4.5, -1.5 / 4.5, -1.5 / 10.5, 1.5 / 10.5]).unsqueeze(-1) * response_mask
```
Comment (SumanthRH) [inline tests/backends/skyrl_train/utils/test_ppo_utils.py]: let's make sure `expected` includes `1e-6` in the denominator

**Comment by tamoghnokandar** [comment]
Done!

File: examples/train/algorithms/maxrl/run_maxrl_gsm8k.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+# You can override the default values with e.g.: `NUM_GPUS=1 bash examples/train/algorithms/maxrl/run_maxrl_gsm8k.sh`.
+
+: "${DATA_DIR:="$HOME/data/gsm8k"}"
+: "${NUM_GPUS:=4}"
+: "${LOGGER:=wandb}" # change to "console" to print to stdout
+: "${INFERENCE_BACKEND:=vllm}"
+
+# MAXRL parameters
+: "${ADV_ESTIMATOR:=maxrl}"
+: "${USE_KL_LOSS:=true}"
```
Comment (SumanthRH) [inline examples/train/algorithms/maxrl/run_maxrl_gsm8k.sh:21]: MaxRL doesn't really recommend/ require KL loss to be true, it is set independently

```suggestion

# Other algorithm parameters
: "${USE_KL_LOSS:=true}"
```

File: tests/tx/models/test_models_common.py
Comment (SumanthRH) [reply tests/tx/models/test_models_common.py:1]: @tamoghnokandar

File: skyrl/backends/skyrl_train/utils/ppo_utils.py [hunk 3]
```diff
@@ -1182,6 +1188,41 @@ def compute_grpo_outcome_advantage(
...
+        for idx in id2score:
+            if len(id2score[idx]) == 1:
+                id2mean[idx] = torch.tensor(0.0)
+            elif len(id2score[idx]) > 1:
+                id2mean[idx] = torch.mean(torch.tensor(id2score[idx]))
+            else:
+                raise ValueError(f"no score in prompt index: {idx}")
+        for i in range(bsz):
+            if len(id2score[index[i]]) > 1:
+                scores[i] = (scores[i] - id2mean[index[i]]) / (id2mean[index[i]] + epsilon)
```
[the remark this answers is not in the document]
Comment (SumanthRH) [reply skyrl/backends/skyrl_train/utils/ppo_utils.py:1270]: This formulation is as per the original maxrl paper

File: tests/tx/models/test_models_common.py
Comment (tamoghnokandar) [reply tests/tx/models/test_models_common.py:1]: I think it is already restored? I have made it same as the current main branch.

File: skyrl/backends/skyrl_train/utils/ppo_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tamoghnokandar) [reply skyrl/backends/skyrl_train/utils/ppo_utils.py:1270]: Should I make the denominator absolute then? Don't think people use negative rewards anyways nowadays

File: tests/tx/models/test_models_common.py
Comment (SumanthRH) [reply tests/tx/models/test_models_common.py:1]: Yeah no issues, I reverted the change in this commit: https://github.com/NovaSky-AI/SkyRL/pull/1126/changes/328943a5ba160690e3bd3cbd09908424dc6fc926
